File: sdk/python/src/capsule_sdk/resources/workloads.py
from __future__ import annotations

from collections.abc import Mapping
from pathlib import Path
from typing import TYPE_CHECKING, Any, cast

import yaml

from capsule_sdk._errors import CapsuleNotFound
from capsule_sdk.models.layered_config import (
    BuildResponse,
    CreateConfigResponse,
    LayeredConfigDetail,
    StoredLayeredConfig,
)
from capsule_sdk.models.workload import ResolvedWorkloadRef, WorkloadSummary

if TYPE_CHECKING:
    from os import PathLike

    from capsule_sdk.models.runner import AllocateRunnerResponse
    from capsule_sdk.resources.layered_configs import LayeredConfigs
    from capsule_sdk.resources.runners import Runners
    from capsule_sdk.runner_config import RunnerConfig
    from capsule_sdk.runner_session import RunnerSession

class Workloads:
    """High-level workload onboarding and runtime API."""

    def __init__(self, layered_configs: LayeredConfigs, runners: Runners) -> None:
        self._layered_configs = layered_configs
        self._runners = runners

    def onboard(
        self,
        spec: RunnerConfig | dict[str, Any] | str | PathLike[str],
        *,
        name: str | None = None,
        build: bool = True,
        force: bool = False,
        clean: bool = False,
    ) -> WorkloadSummary:
        """Register a workload from a RunnerConfig, dict, YAML string, or YAML file."""
        body = self._normalize_spec(spec, name=name)
        created = self._layered_configs.create(body)
        if build:
            self._layered_configs.build(created.config_id, force=force, clean=clean)
        return WorkloadSummary(
            display_name=body["display_name"],
            config_id=created.config_id,
            workload_key=created.leaf_workload_key,
        )

    def onboard_yaml(
        self,
        yaml_spec: str | PathLike[str],
        *,
        name: str | None = None,
        build: bool = True,
        force: bool = False,
        clean: bool = False,
    ) -> WorkloadSummary:
        """Register a workload from YAML text or a YAML file path."""
        return self.onboard(yaml_spec, name=name, build=build, force=force, clean=clean)

    def list(self) -> list[WorkloadSummary]:
        return [self._to_summary(cfg) for cfg in self._layered_configs.list()]

    def _resolve_ref(
        self,
        workload: (
            str
            | WorkloadSummary
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
        ),
    ) -> ResolvedWorkloadRef:
        if isinstance(workload, WorkloadSummary):
            return ResolvedWorkloadRef(
                display_name=workload.display_name,
                config_id=workload.config_id,
                workload_key=workload.workload_key,
            )
        if isinstance(workload, LayeredConfigDetail):
            return ResolvedWorkloadRef(
                display_name=workload.config.display_name,
                config_id=workload.config.config_id,
                workload_key=workload.config.leaf_workload_key,
            )
        if isinstance(workload, StoredLayeredConfig):
            return ResolvedWorkloadRef(
                display_name=workload.display_name,
                config_id=workload.config_id,
                workload_key=workload.leaf_workload_key,
            )
        if isinstance(workload, CreateConfigResponse):
            return ResolvedWorkloadRef(
                config_id=workload.config_id,
                workload_key=workload.leaf_workload_key,
            )
        return self._layered_configs.resolve_workload_ref(workload)

    def get(
        self,
        workload: (
            str
            | WorkloadSummary
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
        ),
    ) -> WorkloadSummary:
        if isinstance(workload, WorkloadSummary):
            return workload
        resolved = self._resolve_ref(workload)
        return self._summary_from_resolved(resolved)

    def build(
        self,
        workload: (
            str
            | WorkloadSummary
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
        ),
        *,
        force: bool = False,
        clean: bool = False,
    ) -> BuildResponse:
        summary = self.get(workload)
        if not summary.config_id:
            raise CapsuleNotFound(f"Workload {summary.display_name!r} does not have a config_id.")
        return self._layered_configs.build(summary.config_id, force=force, clean=clean)

    def delete(
        self,
        workload: (
            str
            | WorkloadSummary
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
        ),
    ) -> None:
        summary = self.get(workload)
        if not summary.config_id:
            raise CapsuleNotFound(f"Workload {summary.display_name!r} does not have a config_id.")
        self._layered_configs.delete(summary.config_id)

    def start(
        self,
        workload: (
            str
            | WorkloadSummary
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
        ),
        **kwargs: Any,
    ) -> RunnerSession:
        resolved = self._resolve_ref(workload)
        return self._runners.from_config(resolved, **kwargs)

    def allocate(
        self,
        workload: (
            str
            | WorkloadSummary
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
        ),
        **kwargs: Any,
    ) -> AllocateRunnerResponse:
        resolved = self._resolve_ref(workload)
        return self._runners.allocate(resolved, **kwargs)

    @staticmethod
    def _to_summary(cfg: StoredLayeredConfig) -> WorkloadSummary:
        return WorkloadSummary(
            display_name=cfg.display_name or cfg.leaf_workload_key or cfg.config_id,
            config_id=cfg.config_id,
            workload_key=cfg.leaf_workload_key,
        )

    @staticmethod
    def _summary_from_resolved(ref: ResolvedWorkloadRef) -> WorkloadSummary:
        display_name = ref.display_name or ref.workload_key or ref.config_id
        if not display_name:
            raise CapsuleNotFound("Resolved workload reference is missing a display name and identifiers.")
        return WorkloadSummary(
            display_name=display_name,
            config_id=ref.config_id,
            workload_key=ref.workload_key,
        )

    def _normalize_spec(
        self,
        spec: RunnerConfig | dict[str, Any] | str | PathLike[str],
        *,
        name: str | None = None,
    ) -> dict[str, Any]:
        if hasattr(spec, "to_create_body"):
            body = cast(dict[str, Any], spec.to_create_body())  # type: ignore[no-any-return]
            return self._ensure_display_name(body, provided_name=name)

        if isinstance(spec, Mapping):
            return self._normalize_mapping(spec, provided_name=name)

        raw_spec = str(spec)
        if "\n" in raw_spec:
            yaml_path = None
            source_name = None
            raw_text = raw_spec
        else:
            yaml_path = Path(raw_spec)
            try:
                exists = yaml_path.exists()
            except OSError:
                exists = False
            source_name = yaml_path.stem if exists else None
            raw_text = yaml_path.read_text() if exists else raw_spec
        loaded = yaml.safe_load(raw_text)
        if not isinstance(loaded, Mapping):
            raise ValueError("Workload YAML must parse to a mapping/object.")
        return self._normalize_mapping(
            cast(Mapping[str, Any], loaded),
            provided_name=name,
            source_name=source_name,
        )

    def _normalize_mapping(
        self,
        spec: Mapping[str, Any],
        *,
        provided_name: str | None = None,
        source_name: str | None = None,
    ) -> dict[str, Any]:
        raw = dict(spec)
        workload_spec = raw.get("workload")
        body: dict[str, Any] = (
            dict(cast(Mapping[str, Any], workload_spec))
            if isinstance(workload_spec, Mapping)
            else raw
        )
        return self._ensure_display_name(body, provided_name=provided_name, source_name=source_name)

    @staticmethod
    def _ensure_display_name(
        body: Mapping[str, Any],
        *,
        provided_name: str | None = None,
        source_name: str | None = None,
    ) -> dict[str, Any]:
        normalized = dict(body)
        display_name = (
            normalized.get("display_name")
            or normalized.get("name")
            or provided_name
            or source_name
        )
        if not isinstance(display_name, str) or not display_name:
            raise ValueError(
                "Workload specs must provide a display name via `display_name`, `name`, or the `name=` argument."
            )
        normalized["display_name"] = display_name
        normalized.pop("name", None)
        return normalized

File: sdk/typescript/src/resources/workloads.ts
import fs from "node:fs";

import YAML from "js-yaml";

import { CapsuleNotFound } from "../errors";
import type {
  BuildResponse,
  CreateConfigResponse,
  LayeredConfigDetail,
  StoredLayeredConfig,
} from "../models/layered-config";
import {
  createResolvedWorkloadRef,
  createWorkloadSummary,
  type ResolvedWorkloadRef,
  type WorkloadSummary,
} from "../models/workload";
import type { RunnerConfig } from "../runner-config";
import type { RunnerSession } from "../runner-session";
import { validateConfigId } from "../validation";
import type { AllocateRunnerOptions, Runners } from "./runners";
import type { LayeredConfigs } from "./layered-configs";

export type WorkloadInput =
  | string
  | WorkloadSummary
  | CreateConfigResponse
  | StoredLayeredConfig
  | LayeredConfigDetail
  | RunnerConfig;

export interface OnboardOptions {
  name?: string;
  build?: boolean;
  force?: boolean;
  clean?: boolean;
}

export class Workloads {
  public constructor(
    private readonly layeredConfigs: LayeredConfigs,
    private readonly runners: Runners,
  ) {}

  public async onboard(
    spec: RunnerConfig | Record<string, unknown> | string,
    options: OnboardOptions = {},
  ): Promise<WorkloadSummary> {
    const body = this.normalizeSpec(spec, options.name);
    const created = await this.layeredConfigs.create(body);

    if (options.build ?? true) {
      await this.layeredConfigs.build(created.config_id, {
        force: options.force,
        clean: options.clean,
      });
    }

    return createWorkloadSummary({
      display_name: String(body.display_name),
      config_id: created.config_id,
      workload_key: created.leaf_workload_key,
    });
  }

  public async onboardYaml(yamlSpec: string, options: OnboardOptions = {}): Promise<WorkloadSummary> {
    return this.onboard(yamlSpec, options);
  }

  public async list(): Promise<WorkloadSummary[]> {
    const configs = await this.layeredConfigs.list();
    return configs.map((cfg) => this.toSummary(cfg));
  }

  public async get(workload: WorkloadInput): Promise<WorkloadSummary> {
    if (isWorkloadSummary(workload)) {
      return workload;
    }

    const resolved = await this.resolveRef(workload);
    return this.summaryFromResolved(resolved);
  }

  public async build(
    workload: WorkloadInput,
    options: { force?: boolean; clean?: boolean } = {},
  ): Promise<BuildResponse> {
    const summary = await this.get(workload);
    if (!summary.config_id) {
      throw new CapsuleNotFound(`Workload ${summary.display_name} does not have a config_id.`);
    }
    return this.layeredConfigs.build(summary.config_id, options);
  }

  public async delete(workload: WorkloadInput): Promise<void> {
    const summary = await this.get(workload);
    if (!summary.config_id) {
      throw new CapsuleNotFound(`Workload ${summary.display_name} does not have a config_id.`);
    }
    await this.layeredConfigs.delete(summary.config_id);
  }

  public async start(workload: WorkloadInput, options: AllocateRunnerOptions & { pollInterval?: number; waitReady?: boolean } = {}): Promise<RunnerSession> {
    const resolved = await this.resolveRef(workload);
    return this.runners.fromConfig(resolved, options);
  }

  public async allocate(workload: WorkloadInput, options: AllocateRunnerOptions = {}) {
    const resolved = await this.resolveRef(workload);
    return this.runners.allocate(resolved, options);
  }

  private async resolveRef(workload: WorkloadInput): Promise<ResolvedWorkloadRef> {
    if (isWorkloadSummary(workload)) {
      return createResolvedWorkloadRef(workload);
    }

    if (typeof workload === "object" && workload !== null) {
      if ("config" in workload) {
        const detail = workload as LayeredConfigDetail;
        return createResolvedWorkloadRef({
          display_name: detail.config.display_name,
          config_id: detail.config.config_id,
          workload_key: detail.config.leaf_workload_key,
        });
      }

      if ("leaf_workload_key" in workload && "config_id" in workload) {
        const config = workload as StoredLayeredConfig | CreateConfigResponse;
        return createResolvedWorkloadRef({
          display_name: "display_name" in config ? (config as StoredLayeredConfig).display_name : undefined,
          config_id: config.config_id,
          workload_key: config.leaf_workload_key,
        });
      }

      if ("workload_key" in workload) {
        return createResolvedWorkloadRef(workload as ResolvedWorkloadRef);
      }
    }

    return this.layeredConfigs.resolveWorkloadRef(workload);
  }

  private toSummary(config: StoredLayeredConfig): WorkloadSummary {
    return createWorkloadSummary({
      display_name: config.display_name ?? config.leaf_workload_key ?? config.config_id,
      config_id: config.config_id,
      workload_key: config.leaf_workload_key,
    });
  }

  private summaryFromResolved(ref: ResolvedWorkloadRef): WorkloadSummary {
    const displayName = ref.display_name ?? ref.workload_key ?? ref.config_id;
    if (!displayName) {
      throw new CapsuleNotFound("Resolved workload reference is missing a display name and identifiers.");
    }

    return createWorkloadSummary({
      display_name: displayName,
      config_id: ref.config_id,
      workload_key: ref.workload_key,
    });
  }

  private normalizeSpec(
    spec: RunnerConfig | Record<string, unknown> | string,
    providedName?: string,
  ): Record<string, unknown> {
    if (typeof spec === "object" && spec !== null && "toCreateBody" in spec && typeof spec.toCreateBody === "function") {
      return this.ensureDisplayName((spec as RunnerConfig).toCreateBody() as unknown as Record<string, unknown>, providedName);
    }

    if (typeof spec === "object" && spec !== null && !Array.isArray(spec) && !("toCreateBody" in spec)) {
      return this.normalizeMapping(spec, providedName);
    }

    const rawSpec = String(spec);
    if (rawSpec.includes("\n")) {
      return this.normalizeMapping(this.parseWorkloadYaml(rawSpec), providedName);
    }

    if (fs.existsSync(rawSpec)) {
      const text = fs.readFileSync(rawSpec, "utf8");
      return this.normalizeMapping(
        this.parseWorkloadYaml(text),
        providedName,
        rawSpec.split("/").pop()?.replace(/\.[^.]+$/u, ""),
      );
    }

    return this.normalizeMapping(this.parseWorkloadYaml(rawSpec), providedName);
  }

  private normalizeMapping(
    spec: Record<string, unknown>,
    providedName?: string,
    sourceName?: string,
  ): Record<string, unknown> {
    const workloadSpec = spec.workload;
    const body =
      workloadSpec && typeof workloadSpec === "object" && !Array.isArray(workloadSpec)
        ? { ...(workloadSpec as Record<string, unknown>) }
        : { ...spec };

    return this.ensureDisplayName(body, providedName, sourceName);
  }

  private ensureDisplayName(
    body: Record<string, unknown>,
    providedName?: string,
    sourceName?: string,
  ): Record<string, unknown> {
    const normalized = { ...body };
    const displayName = normalized.display_name ?? normalized.name ?? providedName ?? sourceName;

    if (typeof displayName !== "string" || displayName.length === 0) {
      throw new Error(
        "Workload specs must provide a config ID (slug) via `display_name`, `name`, or the `name=` argument.",
      );
    }

    validateConfigId(displayName);
    normalized.display_name = displayName;
    delete normalized.name;
    return normalized;
  }

  private parseWorkloadYaml(text: string): Record<string, unknown> {
    const loaded = YAML.load(text);
    if (typeof loaded !== "object" || loaded === null || Array.isArray(loaded)) {
      throw new Error("Workload YAML must parse to a mapping/object.");
    }
    return loaded as Record<string, unknown>;
  }
}

function isWorkloadSummary(value: WorkloadInput): value is WorkloadSummary {
  return typeof value === "object" && value !== null && "display_name" in value && !("config" in value);
}
